In [1]:
def reverse_parameter(param):
    if isinstance(param, str):
        return ''.join(reversed(param))  # Using reversed() and join() for strings
    elif isinstance(param, int):
        is_negative = param < 0
        reversed_num = int(str(abs(param))[::-1])  # Handle the sign separately
        return -reversed_num if is_negative else reversed_num
    return None

# Example usage
a = reverse_parameter("abc")
b = reverse_parameter(123)
c = reverse_parameter(-123)
print(a,b,c)

cba 321 -321


In [3]:
s = 'asdfghjasfgasdfgh'
from collections import Counter
s = Counter(str(s)) 
s


Counter({'a': 3, 's': 3, 'f': 3, 'g': 3, 'd': 2, 'h': 2, 'j': 1})

abstraction - hiding the implementations details of class and only show the essntial features to the user 
Encapsulation -warppiing  data and function in to single unit (object)

In [1]:
# Input: arr[] = [“geeksforgeeks”, “geeks”, “geek”, “geezer”]
# Output: “gee”.

def longestCommonPrefix(arr):
    if not arr:
        return ''
    prefix =arr[0]
    for char in arr[1:]:
        while not char.startswith(prefix):
            prefix =prefix[:-1]
            if not prefix:
                return ""
    return prefix 

arr =["geeksofgeeks","geeksof"]
print(longestCommonPrefix(arr))
    
    


geeksof


In [4]:
# how to reverse in stering 
s = "lkjhgfdsdfghjlkhf"
s1 = ''
for i in s:
    s1 =i + s1 
print(s1)


fhkljhgfdsdfghjkl


In [23]:
def abc(arr):
    prefix =arr[0]
    for char in arr[1:]:
        if char.startswith(prefix):
            prefix =prefix[:-1]
            if not prefix:
                return  ''
    return prefix 
arr = ["flowghjkabs ","flowword"]
print(abc(arr))

flowghjkabs 


In [5]:
def min_distance(S, word1, word2):
    indices_word1 = []
    indices_word2 = []

    for index, word in enumerate(S):
        if word == word1:
            indices_word1.append(index)
        elif word == word2:
            indices_word2.append(index)

    min_dist = float('inf')
    
    for i in indices_word1:
        for j in indices_word2:
            min_dist = min(min_dist, abs(i - j))
    
    return min_dist

# Input data
S = ["the", "quick", "brown", "fox", "quick"]
word1 = "the"
word2 = "fox"

# Output result
result = min_distance(S, word1, word2)
print(result)  # Output: 3


3


In [19]:
num= 5
factorial = 1
for i in range(1, num + 1):
    factorial = factorial * i
print(f"The factorial of {num} is {factorial}")


The factorial of 5 is 120


In [21]:
dict1 = {"a": 1, "b": 2}
dict2 = {"b": 3, "c": 4}
merged_dict = {**dict1, **dict2}
print(merged_dict)

{'a': 1, 'b': 3, 'c': 4}


In [26]:
text = "hello welcome "
char_count = {}
for char in text:
    char_count[char] = char_count.get(char, 0) + 1
print(str(char_count))

{'h': 1, 'e': 3, 'l': 3, 'o': 2, ' ': 2, 'w': 1, 'c': 1, 'm': 1}


In [27]:
def longest_substring_without_repeating(s):
    char_index_map = {}
    start = max_length = 0

    for index, char in enumerate(s):
        if char in char_index_map and char_index_map[char] >= start:
            start = char_index_map[char] + 1
        char_index_map[char] = index
        max_length = max(max_length, index - start + 1)

    return max_length

input_string = "geeksforgeeks"
result = longest_substring_without_repeating(input_string)
print("Length of the longest substring without repeating characters:", result)

Length of the longest substring without repeating characters: 7


In [1]:
# 9. Find the Longest Palindromic Substring

class solved:
    def longestpalindrom(self, s):
        res =''
        resLen = 0 
        for i in range(len(s)):
            # odd length
            l , r = i, i
            while  l>=0 and r < len(s) and s[l] ==s[r]:
                if (r -1 + 1) > resLen :
                    res = s[l: r+1] 
                    resLen = r - l +1
                l -= 1 
                r += 1 
            #even length
            l, r = i, i+1 
            while l >= 0 and r < len(s)  and s[l] == s[r]:
                if (r-l + 1) > resLen:
                    res  =s[l: r+1]
                    resLen  = r -l + 1
                l -= 1
                r += 1

        return res


        

In [ ]:
nested_list = [[1, 2], [3, 4], [5, 6] ,[[1,2],1,1,2,[2,3,4 ,[1]]]]
def flatten_list(nested_list):
    flattened = []
    for item in nested_list:
        if isinstance(item, list):
            flattened.extend(flatten_list(item))
        else:
            flattened.append(item)
    return flattened 
nested_list = [[1, 2], [3, 4], [5, 6], [[1, 2], 1, 1, 2, [2, 3, 4, [1]]]]
flattened_list = flatten_list(nested_list)
print(flattened_list)


In [4]:
s =" python programming  programming"
s1 = {}
for char in s:
    if char != " ":
        if char not in  s1 :
            s1[char] = 1
        s1[char] += 1
print(s1) 
o = ' , '.join(f"{char} : {count}" for char,count in s1.items())
print(o)


{'p': 4, 'y': 2, 't': 2, 'h': 2, 'o': 4, 'n': 4, 'r': 5, 'g': 5, 'a': 3, 'm': 5, 'i': 3}
p : 4 , y : 2 , t : 2 , h : 2 , o : 4 , n : 4 , r : 5 , g : 5 , a : 3 , m : 5 , i : 3


In [5]:
# write a program to 100 to 200 find a prime nums
l1 =[]
for num in range(100,201):
    if num > 1 :
        for i in range(2,num):
            if num % i == 0:
                break 
        else:
            l1.append(num)
print(l1)

[101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199]


In [12]:
def tree(hight):
    lenght = hight * 2 - 1
    start = 1
    for  i in range(1, hight+1):
        print(("*" * start).center(lenght))
        start +=2 
    # print("*" .center(lenght))

tree(6)


     *     
    ***    
   *****   
  *******  
 ********* 
***********


In [18]:
l1 = [10,8,2,4,6,4]
def sort_list(l1):
    n = len(l1)
    for i in range(n):
        for j in range(0,n-i-1):
            if l1[j] > l1[j+1]:
                l1[j],l1[j+1] = l1[j+1] ,l1[j] 
    return l1
l1 = [10,8,2,4,6,4]

a = sort_list(l1)
print(a)

[2, 4, 4, 6, 8, 10]


In [21]:
# Fibonacci Series 
def fib_fun(n):
    fib_ser =[]
    a, b = 0, 1
    for i in range(n):
        fib_ser.append(a)
        a, b = b, b+a 
    return fib_ser  
print(fib_fun(15))


[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377]


In [34]:
# 1 first way  reverse to list 
def reverse_fun(l1):
    l2 = [] 
    for i in range(len(l1)-1,-1,-1):
        l2.append(l1[i])
    return l2
l1 =[1, 2, 3, 4, 5]
print(reverse_fun(l1))

# 2nd way
def rev_fun(l1):
    start, end = 0, len(l1)-1
    while start < end :
        l1[start], l1[end] = l1[end], l1[start]
        start += 1
        end -= 1
    return l1
print(rev_fun(l1)) 

l1 =[1,5,3,2,5,7,4,3]
l3= []
for i in l1:
    l3 = [i] + l3
print(l3)

[5, 4, 3, 2, 1]
[5, 4, 3, 2, 1]
[3, 4, 7, 5, 2, 3, 5, 1]


In [42]:
# print the set of duplicates in a list: 
def duplicate_fun(l1):
    not_duplicate = [] 
    deplicate = []
    for i in l1:
        if i in not_duplicate:
            if i not in deplicate:
                deplicate.append(i) 
        else:
            not_duplicate.append(i)
    return deplicate  
l1 =[1,2,3,4,5,56,3,2,21,]
print(duplicate_fun(l1))    


[3, 2]


In [ ]:
# Binary Search Implementation:
def binary_search(l1, x):
    low = 0
    hight = len(l1)-1
    while  low <= hight:
        mid = (low + hight) // 2 
        if l1[mid] == x:
            return mid 
        elif arr[mid] < x:
            low = mid + 1
        hight = mid -1
    return 'not found'
l1 = [10, 20, 30, 40, 50, 60, 70, 80, 90]
x = 50
print(binary_search(l1, x)) 


4


In [44]:
def delete_rec_char(str1):
    see = {}
    out =  ''
    for char in str1:
        if char   not in see :
            see[char] = True
            out += char
    return out 

str1 = "programming"
result = delete_rec_char(str1)
print(result)


        

progamin


In [87]:
x = 'malayalam'
w = ''
for i in x:
    w = i + w  

if x == w:
    print('Palindrome')
else:
    print('Not palindrome')

Palindrome


In [ ]:
# Linear Search in Python

def liner_search(arr, x):
    for i in range(len(arr)):
        if arr[i] == x:
            return i
    return -1

arr =[10,20,30,40,50,60]
x = 40 
res =liner_search(arr, x) 
if res  != 1:
    print(f" element  {x}  found at index {res} .")
else:
    print(f" element {x} not fount in the array .")

 element  40  found at index 3 .


In [5]:
import re 
name = 'Python is 1'

digit = re.sub("[^0-9]","",name) 
latercount =re.sub("[^a-zA-Z]","", name)
spaceCount =re.findall("[ \n]",name)
print(len(digit),len(latercount),len(spaceCount))

1 8 2


In [18]:
import re  
spChar = "!@#$%^&*()" 
count = re.sub('[/w]+', '', spChar)
print(len(count))

10


In [21]:
import re 
string = "C O D E"
spaces = re.compile(r'\s+')
res = re.sub(spaces,"", string)
print(res)

CODE


In [24]:
# Longest Palindromic Substring 
def  longest_substing (s):
    if not s:
        return " " 
    def  expand_around_center(left ,right):
        while left >= 0 and right < len(s) and s[left] == s[right] :
            left -= 1
            right += 1
        return s[left + 1:right] 
    

leg = longest_substing ('babad')
print(leg)


None


In [26]:
def longest_palindromic_substring(s: str) -> str:
    if not s:
        return ""
    
    def expand_around_center(left: int, right: int) -> str:
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        # Return the palindrome substring
        return s[left + 1:right] 
    longest = " " 
    for i in  range(len(s)):
        old_p = expand_around_center(i, i)
        if len(old_p) > len(longest):
            longest =old_p 
        eve_p = expand_around_center(i, i+1)
        if len(eve_p) > len(longest):
            longest = eve_p 
    return longest

leg = longest_palindromic_substring('babad')
print(leg)


None


In [27]:
# . Count and Group Anagrams from a List of Strings

from  collections  import defaultdict 

def  group_anagrams(words: list):
    anagram = defaultdict(list)
    for word in words:
        sorted_words = ''.join(sorted(word))
        anagram[sorted_words] .append(word)
    return dict(anagram) 
print(group_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))

{'aet': ['eat', 'tea', 'ate'], 'ant': ['tan', 'nat'], 'abt': ['bat']}


In [28]:
def full_justify(words, maxWidth):
    result = []
    current_line = []
    current_length = 0

    for word in words:
        if current_length + len(word) + len(current_line) > maxWidth:
            for i in range(maxWidth - current_length):
                current_line[i % (len(current_line) - 1 or 1)] += " "
            result.append("".join(current_line))
            current_line = []
            current_length = 0
        current_line.append(word)
        current_length += len(word)
    
    result.append(" ".join(current_line).ljust(maxWidth))
    
    return result

words = ["This", "is", "an", "example", "of", "text", "justification."]
maxWidth = 16

output = full_justify(words, maxWidth)

for line in output:
    print(f'"{line}"')


"This    is    an"
"example  of text"
"justification.  "


In [32]:
def add (a,b):
    c = a + b
    return c 

res = add(add(5,add(5,6.5)),5),5
print(res)

26.5


In [4]:
def addop():
    a = float(input("Enter First Value:"))
    b = float(input("Enter Second Value:"))
    c= a + b
    print("Sum({},{}) = {}" .format( a, b, c ))

addop()


Sum(19.0,19.0) = 38.0


In [8]:
l = ['ab', 'cd']
output = [{char for char in s[::-1]} for s in l]
print(output)

output =[]
for s in l:
    output_set = set()
    for char in s[::-1]:
        output_set.add(char)
    output.append(output_set) 
output



[{'a', 'b'}, {'d', 'c'}]


[{'a', 'b'}, {'c', 'd'}]

In [18]:
s='aabbaaccaccab'
# output =6a3b4c

out = ''
dic ={}
for char in s:
    dic[char] =dic.get(char,0) + 1
print(dic)
for k,v in dic.items():
    out +=  str(v)+k
print(out)


{'a': 6, 'b': 3, 'c': 4}
6a3b4c


In [25]:
s ='assaddsrdfcx' 
# out = ''
dic = {}
for i in s:
    dic[i] = dic.get( i, 0) + 1
print(dic) 

for k,v in sorted(dic.items()):
    out = k ,' item  ', v ,'times .'
    print(out)
    

{'a': 2, 's': 3, 'd': 3, 'r': 1, 'f': 1, 'c': 1, 'x': 1}
('a', ' item  ', 2, 'times .')
('c', ' item  ', 1, 'times .')
('d', ' item  ', 3, 'times .')
('f', ' item  ', 1, 'times .')
('r', ' item  ', 1, 'times .')
('s', ' item  ', 3, 'times .')
('x', ' item  ', 1, 'times .')


In [ ]:
x = {1: 2, 3: 4, 4: 3, 2: 1, 0: 0}


In [37]:
# 2*2  and 2 to 2000  

# def even_num_squ(nums):
#     for  i in range(2,nums):
#         if i % 2 ==0:
#             print(i ** 2)  


l1 = [i ** 2 for i in range(2, 21) if i % 2 == 0]
l1

# even_num_squ(21)



[4, 16, 36, 64, 100, 144, 196, 256, 324, 400]

In [35]:
def fun(n):
    if n <= 0:
        return "error"
    elif n == 1:
        return 0
    elif n == 2:
        return 1

    return fun(n-1) + fun(n-2) 

In [34]:
# linklist 

class Node :
    def __init__(self,data):
        self.data = data 
        self.next =None 


class Linkedlist :
    def __init__(self):
        self.head = None 
    def append(self, data ):
        new_node =Node(data)
        if not self.head:
            self.head =new_node 
            return 
        last  = self.head
        while  last.head :
            last =last.next 
        last.next =new_node 
    def  display(self):
        current = self.head
        while current :
            print(current.data)
            current = current.next 
        print("None") 


ll =Linkedlist()
l1.append(1)
l1.append(2)
l1.append(3)
ll.display()


        

None


pymalloc
diff between init() and  new()
threadsafe
profiling
how many types of profiling available ?
detached head
~
 
COPY vs ADD  command in docker

check in server application is running 
and last log file  100 line to show

 

In [ ]:
# Find the missing number in a list of consecutive integers.
def find_missing_numbers(l1):
    min_num = min(l1)
    max_num = max(l1)
    full_set = set(range(min_num, max_num + 1))
    input_set = set(l1)
    missing_numbers = sorted(full_set - input_set)
    return missing_numbers

l1 = [1, 2, 4, 7]
print(find_missing_numbers(l1))  
# Output: [3, 5, 6]

In [ ]:
# l1 =[1,2,4,5,8]

def find_missing_num(l1):
    max_num = max(l1)
    min_num = min (l1)
    num =set(range(min_num, max_num +1))
    input_set =set(l1)
    miss_num = num - input_set
    return miss_num
l1 = [1,2,4,5,8]
print(find_missing_num(l1)) 



def funsum(num):
    while num >= 10:
        num = sum(int(digit) for digit in str(num))  
    return num  

print(funsum(567))  



In [ ]:
https://app.zoom.us/wc/81445996222/join?fromPWA=1&pwd=QUzrdz23KwmAaMbDXKvSULg8aYrMjg.1&_x_zm_rtaid=ik0p4SCVQHG1dUSy1uX-ZQ.1738150758263.5a47951bfcce7d

In [1]:
def rev_int_rev(n):
    rev =0
    while  n > 0:
        digit = n % 10
        rev = rev *10 + digit
        n = n // 10 
    return rev 

print(rev_int_rev(123456))

654321


In [6]:
# convert ['a', 'b'] to   "'a' 'b'"
lst = ['a', 'b']
result = ' '.join(f"'{item}'" for item in lst)
print(result)
 

'a' 'b'


Feature	                raise	                                      assert
Purpose	                Manually raise exceptions	                  Debugging & checking conditions
When Used	            Error handling & validation	                  Testing assumptions in development
If Condition Fails	   Always raises an exception                     Raises AssertionError only in debugging mode (disabled with -O flag)

In [2]:
# merge  no and how  to uesd in program
import pandas as pd

df1 = pd.DataFrame({'ID': [1, 2, 3], 'Name': ['Alice', 'Bob', 'Charlie']})
df2 = pd.DataFrame({'ID': [1, 2, 4], 'Score': [85, 90, 78]})

# INNER JOIN (default)
merged_df = pd.merge(df1, df2, on='ID', how='inner')
print(merged_df)

   ID   Name  Score
0   1  Alice     85
1   2    Bob     90


In [5]:
# concat() in pandas 
df1 = pd.DataFrame({'ID': [1, 2], 'Name': ['Alice', 'Bob']})
df2 = pd.DataFrame({'ID': [3, 4], 'Name': ['Charlie', 'David']})

concat_df = pd.concat([df1, df2], axis=0)
print(concat_df)

   ID     Name
0   1    Alice
1   2      Bob
0   3  Charlie
1   4    David


Feature	                      merge()	                                      concat()
Purpose	                    SQL-like joins on columns	                   Stack DataFrames (rows/columns)
Key Requirement	            Needs a common column (key)	                   No common column needed
Join Type	               Supports inner, outer, left, right joins	       Simple concatenation
Default Behavior	       inner join	                                    Vertical stacking (axis=0)

In [11]:
l1 = ['a','b']
res = ''.join(f"'{item}'" for item in l1) 
res


"'a''b'"

In [1]:
p =10000
for i in range(1,10):
    p = p - 0.1 * p 
    print(round(p),end = " ")

9000 8100 7290 6561 5905 5314 4783 4305 3874 

In [3]:
for i in range(1,5):
    for j in range(1,5):
        if  i != j :
            print( i, j)

1 2
1 3
1 4
2 1
2 3
2 4
3 1
3 2
3 4
4 1
4 2
4 3


In [ ]:
x = int(input("enter the number :"))
new = []
if  x == 1:
    print("1 is not prime number ")
else:
    for i  in range(1,x+1):
        if x % i == 0 :
            new.append(i)
    if  len(new) > 2 :
        print("it  is not a prime number ")
    else:
        print("it prime number ")
          

In [7]:
n = int(input("enter the  number :"))
num = 1 
for i in range(0,n):
    for j in range(0, i+1):
        print(num ,end= " ") 
        num = num + 1 
    print( )

for i in range(1,n+1):
    for j in range(1,i+1):
        print(j,end= " ")
    print("\n")

1 
2 3 
4 5 6 
7 8 9 10 
11 12 13 14 15 
1 

1 2 

1 2 3 

1 2 3 4 

1 2 3 4 5 



In [9]:
for i in range(1,11):
    for j in range(1,11):
        print( i * j ,end= " ")
    print("\n")

1 2 3 4 5 6 7 8 9 10 

2 4 6 8 10 12 14 16 18 20 

3 6 9 12 15 18 21 24 27 30 

4 8 12 16 20 24 28 32 36 40 

5 10 15 20 25 30 35 40 45 50 

6 12 18 24 30 36 42 48 54 60 

7 14 21 28 35 42 49 56 63 70 

8 16 24 32 40 48 56 64 72 80 

9 18 27 36 45 54 63 72 81 90 

10 20 30 40 50 60 70 80 90 100 



In [10]:
for i in range(n,0,-1):
    for j in range(i,0,-1):
        print(j,end= " ")
    print("\n")

7 6 5 4 3 2 1 

6 5 4 3 2 1 

5 4 3 2 1 

4 3 2 1 

3 2 1 

2 1 

1 



In [12]:
from itertools import permutations

def is_consonant(letter):
    """Check if a letter is a consonant."""
    return letter.lower() not in 'aeiou'

def count_distinct_words(word):
    unique_letters = set(word)
    
    all_permutations = set(permutations(word))
    
    valid_words = [
        ''.join(p) for p in all_permutations if is_consonant(p[0])
    ]
    
    return len(valid_words)

# Input
input_word = input("Enter a word: ")
# Output
output_count = count_distinct_words(input_word)
print(output_count)


5040


In [ ]:
docker run -d - p 900:800 nginx:1.23  
docker run start 
docker stop (id)
dpcker ps 
doker logs (id )
docker ps -a 


In [ ]:
x = int(input("enter the number :"))
new = []
if x == 1 :
    print("1 is not prime number")
else:
    for i in range(1,x+1):
        if x % i == 0:
            new.append(i)
    if len(new) > 2:
        print("it is not prime number")
    else:
        print("it is prime number .") 
  

In [3]:
n  = int(input("enter the number : "))
for i in range(0,n+1):
    for j in range(0,i):
        print(i,end=" ")
    print("\n")






1 

2 2 

3 3 3 

4 4 4 4 

5 5 5 5 5 



In [4]:
n= input("enter the number : ")
x =list(n)
count = 0
while count <len(x):
    for i in x:
        count += 1
print(count)

3


In [1]:
import pandas as pd 
data = {"Name" :["alice","Bob"], "Age":[25,30]}
df  =pd.DataFrame(data)
df

Name  Age
0  alice   25
1    Bob   30

In [2]:
df.isnull().sum()


Name    0
Age     0
dtype: int64

In [ ]:
# how to fill missing value in DataFrame 
df.fillna(value=0,inplace=True)
df.ffill() # Forword Fill
df.bfill()  # Backword fill 
# how to select a column in pandas 
df["column_name"]
df[["column1"],]


In [ ]:
# 1. Given a word, print the number distinct words, with or without meaning, that can be formed with the letters of word? 
# Each word should start with a consonant.
# Input: word (size of the word can be from 1 to 20)
# Output: number of distinct words

# Test Case 1:
# Input: CAT
# Output: 4
# Explanation: CAT, CTA, TAC, TCA. (ACT, ATC are not valid words, as they start with vowel letter)

# Test Case 2:
# Input: BOB
# Output: 2
# Explanation: BOB, BBO

# Test Case 3:
# Input: CHAIR
# Output: 72

from itertools import permutations 

def func(word):
    vowels = {'a', 'e', 'i', 'o', 'u'}
    unique_permutations = set(''.join(p) for p in permutations(word))
    print(unique_permutations)
    consonant_permutations = [w for w in unique_permutations if w[0].lower() not in vowels] 
    print(consonant_permutations)
    return len(consonant_permutations)


word ='CAT'
print(func(word))


In [ ]:
# . In how many ways you can pay N rupees with 1 rupee, 2 rupee & 5 rupee denominations, In such a way that number of 1 rupee coins
#  always greater than that of 2 rupee coins and number of 2 rupee coins always greater than that of 5 rupee coins. At least one coin 
# should be given from each denomination.



def count_ways(N, x=0, y=0, z=1):
    # Base case: If the sum is exactly N and conditions hold
    if x + 2*y + 5*z == N and x > y > z >= 1:
        return 1
    
    # If the sum exceeds N, return 0
    if x + 2*y + 5*z > N:
        return 0

    # Try increasing the number of coins in a valid order
    count = 0
    if y > z:  # Ensure y > z before increasing x
        count += count_ways(N, x+1, y, z)
    if x > y:  # Ensure x > y before increasing y
        count += count_ways(N, x, y+1, z)
    count += count_ways(N, x, y, z+1)  # Always try increasing z

    return count

N = 10  # Change N to test different amounts
print(count_ways(N))  # Output: Number of ways


In [14]:
s='aabbaaccaccab'
# output =6a3b4c 

d = {}
out =''
for char in s:
    d[char] =d.get(char,0) +1 
print(d)
for k,v in d.items():
    out =out + str(v) + k 
print(out)



{'a': 6, 'b': 3, 'c': 4}
6a3b4c


In [ ]:
# Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target. 
def  two_sum(nums, target):
    hashmap = {}
    for i ,num in enumerate(nums):
        diff = target -num 
        if diff in hashmap:
            return [hashmap[diff], i ]
        hashmap[num] = i 
    return []
        

In [ ]:
# Problem: Given the head of a singly linked list, reverse the list and return the new head. 

class ListNode:
    def __init__(self,val = 0, naxt =None):
        self.val = val
        self.next = next 
    
    def reverse_list(head):
        prev ,curr  = None, head
        while curr :
            temp = curr.next
            curr.next = prev 
            prev = curr 
            curr = temp 
        return prev 
    


In [4]:
# Problem: Given a string s containing {}, (), [], determine if it is valid.

def is_valid(s):
    stack = [] 
    mapping ={ ')' : '(', 
                '}': '{',
                ']' :'[' }
    for char in s:
        if char in mapping:
            top_ele =stack.pop() if stack else '#'
            if mapping[char] != top_ele:
                return False 
        else:
            stack.append(char)
    return not stack 

    



In [7]:
# how to connect in database in python 
# ! pip install mysql-connector-python

import mysql.connector

conn = mysql.connector.connect(
    host_name =  "localhost"
    user = "user_name"
    password ="password!@321"
    database = "database name "
)

# create cursor
curser =conn.cursor() 

curser.execute("select database();") 


curser.close()
conn.close()



import psycopg2

# Establish connection
conn = psycopg2.connect(
    host="localhost",
    database="your_database",
    user="your_username",
    password="your_password"
)

# Create a cursor
cursor = conn.cursor()

# Execute a query
cursor.execute("SELECT version();")

# Fetch and print results
db_version = cursor.fetchone()
print("Connected to:", db_version)

# Close the connection
cursor.close()
conn.close()



SyntaxError: invalid syntax. Perhaps you forgot a comma? (2242485365.py, line 7)

In [ ]:
# 1. Using Python with Pandas for Batch Processing 
import pandas as pd  
check_size =10000

for chunk in pd.read_csv('learge_data.csv',checksize =check_size):
    chunk['processed_col'] =chunk[row_col] * 2 
    chunk.to_csv("processed_data.csv",mode ='a',header= False, index =False) 
    



In [ ]:
from pyspark.sql import SparkSession

# Initialize Spark Session
spark = SparkSession.builder.appName("BatchProcessing").getOrCreate()

# Read large dataset
df = spark.read.csv("large_data.csv", header=True, inferSchema=True)

# Perform transformations
df = df.withColumn("processed_col", df["raw_col"] * 2)

# Save the processed data
df.write.csv("processed_output", mode="overwrite", header=True)


In [6]:
def freq_word():
    str = input("enter the  string : ")
    l1 = str.split()
    d = {}

    for i in l1 :
        d[i] = d.get(i,0) +1
        # if i not in d.keys():
        #     d[i] = 0
        # d[i] =d[i]+ 1
    print(d) 
freq_word() 






{'hello': 8, 'is': 8, 'my': 4, 'name': 4}


In [1]:
def student_total_marks(N , M , marks):
    subject_averages = [sum (marks[i][j] for i in range(N)) / N for j in range(M)] 
    
    min_avg_sub =subject_averages.index(min(subject_averages))
    print(min_avg_sub)
    result = [sum(student) -student[min_avg_sub] for student in marks] 
    return result 

N = 3
M = 4 

marks = [[80,90,85,70],
        [75,85,80,65],
        [60,75,70,55] ] 

print(student_total_marks( N, M, marks) )



3
[255, 240, 205]


In [3]:
def max_distinct_char_between_y(X,Y,):
    indices = [i for i ,c  in enumerate(X) if c == Y ]
    if len(indices) < 2:
        return -1 
    
    max_distance = 0 
    for i in range(len(indices)):
        for j in range(i+1, len(indices)):
            sub_str = X[indices[i] +1 : indices[j]]
            filtered_chars = set(c for c in sub_str if c != ' ' and c != Y)
            max_distance = max(max_distance ,len(filtered_chars))
    return max_distance

X =" abc  aabc  bcadab"
Y ="a"
print(max_distinct_char_between_y(X,Y))


3


In [ ]:
l1 = [2, 1, 3, 4, 5, 5, 5]
first_max = second_max = float('-inf')

for number in l1:
    if number > first_max:
        second_max = first_max
        first_max = number
    elif first_max > number > second_max:
        second_max = number

print(second_max)  # Output: 4

In [ ]:
str1 = 'abc#$_bcd' 
# out =dcb#$_cba

def revr_fun(str1):
    stack  = [char for char in str1 if char.isalpha()] 
    print(stack)   #  ['a', 'b', 'c', 'b', 'c', 'd'] 
    
    result = []
    for ch in str1:
        if ch.isalpha():
            result.append(stack.pop())
        else:
            result.append(ch)
    print(result) 
    return ''.join(result)

revr_fun(str1)



['a', 'b', 'c', 'b', 'c', 'd']
['d', 'c', 'b', '#', '$', '_', 'c', 'b', 'a']


'dcb#$_cba'

In [7]:
str1 = 'abchd' 
str2 = 'abdf'
# out = abd 
res =''
for char in str1:
    if char in str2:
        res = res + char
print(res)

abd


In [ ]:
s = 'python programming programming' 

dup ={}
for char in s:
    if char in l1:
        dup[char] = dup.get(char, 1) + 1 
    else:
        l1.append(char)
print(dup)


{'p': 3, 'o': 3, 'r': 4, 'm': 4, 'n': 3, 'g': 4, ' ': 2, 'a': 2, 'i': 2}
['p', 'y', 't', 'h', 'o', 'n', ' ', 'r', 'g', 'a', 'm', 'i']


In [ ]:
# prime num 

low =200
upper =500
for num in range(low, upper+1):
	if num>1:
		for i in range (2,num):
			if (num % i)== 0:
				break
		else:
			print(num)

In [4]:
num = 123456
rev =0
while num !=0:
    rev = rev * 10 + num % 10
    num //=10 
print(rev)

654321


In [3]:
num = 8
def fib_fun(num):
    if num <=1:
        return num 
    else:
        return (fib_fun(num-1) + (fib_fun(num-2))) 
    
print(fib_fun(num))


21


In [1]:
s ='ababacabcbxabcde' 
def lengest_sub_string(s):
    char_set =set()
    l =0
    res = 0
    for  r in range(len(s)):
        while s[r] in char_set:
            char_set.remove(s[r])
            l+=1
        char_set.add(s[r])
        res =max(res,r-l+1)
    return res 
print(lengest_sub_string(s))


6


In [34]:
def digital_root(n):
    while n >= 10:  
        n = sum(int(digit) for digit in str(n))  
    return n

n = 567
print(digital_root(n))

# n =567
# 4+6+7 =18 
# 1+8 =9  
# out =9

9


In [35]:
l1 =[1,0,2,0,3,0,4,0]
for i in l1:
    if i == 0:
        l1.remove(i)
        l1.append(i)
print(l1)


[1, 2, 3, 4, 0, 0, 0, 0]


Coding Questions

1. Given a word, print the number distinct words, with or without meaning, that can be formed with the letters of word? Each word should start with a consonant.
Input: word (size of the word can be from 1 to 20)
Output: number of distinct words

Test Case 1:
Input: CAT
Output: 4
Explanation: CAT, CTA, TAC, TCA. (ACT, ATC are not valid words, as they start with vowel letter)

Test Case 2:
Input: BOB
Output: 2
Explanation: BOB, BBO

Test Case 3:
Input: CHAIR
Output: 72

Other testcases:
Input            Output
1. A                   0
2. N                   1
3. OO                0   
4. TO                1
5. MT               2
6. BAT             4
7. CRY             6
8. SUPER         72
9. INDIA          24
10. BUTTER    240




from itertools import permutations 
def count_consonant_start_permutations(word):
 vowels = {'a', 'e', 'i', 'o', 'u'} 
unique_permutations = set(''.join(p) for p in permutations(word)) # Filter words starting with consonants 
consonant_permutations = [w for w in unique_permutations if w[0].lower() not in vowels] 
return len(consonant_permutations)



2. In how many ways you can pay N rupees with 1 rupee, 2 rupee & 5 rupee denominations, In such a way that number of 1 rupee coins always greater than that of 2 rupee coins and number of 2 rupee coins always greater than that of 5 rupee coins. At least one coin should be given from each denomination.

Input: N – the amount in rupees (1 <= N <= 100)
Output: the number of ways in which you can pay the amount as described

Test case 1:
Input: 10
Output: 0

Test case 2:
Input: 15
Output: 2      
Explanation: Possible ways in which you can pay 15 rupees are (1 5rs, 2 2rs, 6 1rs) and (1 5rs, 3 2rps, 4 1rps).


Other testcases:
Input            Output [5rs, 2rs, 1rs]
1. 1                    0
2. 11                  0
3. 12                  1    (1, 2, 3)
4. 14                  1    (1, 2, 5)            
5. 20                  4    (1, 2, 11), (1, 3, 9), (1, 4, 7) (2, 3, 4)
6. 21                  5    (1, 2, 12), (1, 3, 10), (1, 4, 8), (1, 5, 6), (2, 3, 5)
7. 30                  12    (1, 2, 21), (1, 3, 19), (1, 4, 17), (1, 5, 15), (1, 6, 13), (1, 7, 11), (1, 8, 9), (2, 3, 14), (2, 4, 12), (2, 5, 10), (2, 6, 8), (3, 4, 7)    









3. A robbery team went to a city to rob. Every house in the city has a number lock with different number of rings. Insider information tells them that every key combination is a palindrome. ex: 121, 4554. 

 
How many different combinations they have to check for a given lock with n rows? For even number of rows, the key starts with an even number & for an odd number of rows, the key starts with odd number.

Input: n – the number of rings (1 <= n <= 30)
Output: number of keys they have to verify

Test Case 1:
Input: 2
Output: 5 
Explanation: as n is even, the key should start with an even number and it should be a palindrome. So, the possible keys are 00, 22, 44, 66 and 88.

Test Case 2:
Input: 3
Output: 50
Explanation: as n is even, the key should start with an odd number and it should be a palindrome. So, the possible keys are 1n1, 3n3, 5n5, 7n7, 9n9 where n can be from 0 to 9 i.e., 10. So, total 5 times 10 is 50.




Other testcases:
Input            Output
1. 1                  5
2. 4                  50
3. 5                  500   
4. 6                  500
5. 10                50000   
6. 20                5000000000  
7. 25                5000000000000               
8. 30                500000000000000




4. There is test for 100 marks. The test consists of 1mark, 2mark, 5 marks & 10 marks questions as below:
marks/question   X No. of questions      = sub-total marks
10                          X  3                           =  30
5                            X  8                           =  40
2                           X  10                          =  20
1                           X  10                          =  10
----------------------------------------------------------
Total                                                        =  100
----------------------------------------------------------

There are no partial marks given to any answer. Student will get either full marks or zero. No negative marks for the incorrect answers. 

Given X, print in how many possible ways a student can get X number of marks.

Input: X – number of marks (0 <= X <= 100)
Output: Number of ways in which the student can score X marks

Test Case 1:
Input: 2
Output: 55
Explanation: 
Student might have answered two 1-mark questions correctly (any 2 out of 10 i.e., in 45 ways) 
Or student might have answered one 2-mark questions correctly (any 1 out of 10 i.e., in 10 ways)
So, total 45 + 10 = 55

Test Case 2:
Input: 3
Output: 220
Explanation: 
Student might have answered three 1-mark questions correctly (any 3 out of 10 i.e., in 120 ways) 
Or student might have answered one 2-mark question & one 1-mark questions correctly (any 1 out of 10 i.e., in 10 ways * any 1 out of 10 in 10 ways i.e., total 100 ways)
So, total 120 + 100 = 220


Other testcases:
Input            Output
1. 0                   1
2. 1                   10
3. 4                   705   
4. 5                   1660
5. 97                 220
6. 96                 705
7. 98                 55
8. 99               10
9. 100             1


********************************************************************************************
5. Given an n X n plane, how many ways a person can travel from starting cell (yellow) to the destination cell (green) without stepping into the danger cell(red). There is only one danger cell which is always below the destination cell. The person can move only 1 step at a time and can only move to right or up direction.

 

Input: n - size of the plan (2 <= n <= 20)
Output: number of ways the person can travel

Test Case 1:
Input: 3
Output: 3

Test Case 2:
Input: 4
Output: 7

Other testcases:
Input            Output
1. 2                   1
2. 4                   10
3. 5                   35
3. 6                   126
3. 7                   462
4. 8                   1716
5. 10                 24310



In [2]:
arr =[1,9,8,7,6,5,4,3,2,5,4,3,2]
n =len(arr)
for i in range(n):
    for j in range(i+1,n):
        if arr[i] > arr[j]:
            arr[i],arr[j] = arr[j],arr[i]
print(arr)

[1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 7, 8, 9]


In [10]:
import pandas as pd
import pandas as pd

s1 = pd.Series([1, 2, 3], index=['a', 'b', 'c'])
s2 = pd.Series([4, 5, 6], index=['d', 'e', 'f'])

df = pd.DataFrame({'s1': s1, 's2': s2})

s1 = pd.Series([1, 2, 3], index=['a', 'b', 'c'])
s2 = pd.Series([4, 5, 6], index=['d', 'e', 'f'])  

df = pd.merge(s1.to_frame('s1'), s2.to_frame('s2'), left_index=True, right_index=True)
df

Empty DataFrame
Columns: [s1, s2]
Index: []

In [16]:
names = ["Alice", "Bob", "Charlie"] 
ages = [25, 30, 35] 
res = tuple(zip(names,ages))
print(res)


(('Alice', 25), ('Bob', 30), ('Charlie', 35))


Given an array of positive integers nums and a positive integer target, return the minimal length of a subarray whose sum is greater than or equal to target. If there is no such subarray, return 0 instead.

 

Example 1:
Input: target = 7, nums = [2,3,1,2,4,3]
Output: 2
Explanation: The subarray [4,3] has the minimal length under the problem constraint.


Example 4:
Input: target = 8, nums = [4,4]
Output: 2


Example 2:
Input: target = 4, nums = [1,4,4]
Output: 1

Example 3:
Input: target = 11, nums = [1,1,1,1,1,1,1,1]
Output: 0
 

Constraints:

1 <= target <= 109
1 <= nums.length <= 10^5
1 <= nums[i] <= 10^4


In [ ]:
#reverse a string preserving space index

# str1 = "i am doing coding in python"
# Output = "n oh typni gnidoc gn iodmai"

def  reverse_fun(s):
    char_list = list(s)
    rev = [c for c in  char_list if c != ''][::-1]
    print(rev)
    result = []
    j  = 0
    for  ch in char_list:
        if ch ==' ':
            result.append(' ')
        else:
            result.append(rev[j])
            j += 1
    return ''.join(result)
s = "i am doing coding in python"

reverse_fun(s)



['n', 'o', 'h', 't', 'y', 'p', ' ', 'n', 'i', ' ', 'g', 'n', 'i', 'd', 'o', 'c', ' ', 'g', 'n', 'i', 'o', 'd', ' ', 'm', 'a', ' ', 'i']


'n oh typ n i gnid oc  gniod'

In [8]:
def pyramid_fun(n):
    for i in range(1,n+1):
        for j in range(n-1):
            print(" ", end ='')
        for k in range(1, 2*i):
            print("*",end =' ')
        print()


pyramid_fun(5)

    * 
    * * * 
    * * * * * 
    * * * * * * * 
    * * * * * * * * * 


In [ ]:
def abc_fun(input_str):
    result= []
    i = 0
    while i <len(input_str):
        char = input_str[i]
        
        if i + 1 < len(input_str) and input_str[i + 1].isdigit():
            repeat_count = int(input_str[i + 1])
            result.append(char * repeat_count )
            i += 2
        else:
            result.append(char)
            i += 1
    return ''.join(result)

input_str = 'a2bc5d'
print(abc_fun(input_str))

In [ ]:
# 1 - prime num 1 to 100 
# 2 - find factral num 
# 3 - fibenic serise
# 4 - any area for trangle 

# prime nums 
for  num in range(1 ,101):
    if num > 1 :
        for i in range(2, num):
            if num% i ==0:
                break
        else:
            print(num)
 

def fact_num(n):
    if n==0 or n ==1:
        return 1 
    return n * fact_num(n-1) 
    
print(fact_num(5))


def fib_fun(n):
    if n<= 1 :
        return n 
    else:
        return fib_fun(n-1) + fib_fun(n-2)


for i in range(10):
    print(fib_fun(i))




def trangle_fun(base, hight):
    return 0.5 *base * hight

print(trangle_fun(4,5))







In [ ]:
import pandas as pd

class Person:
    """Represents a person with a name, age, and city."""
    def __init__(self, name, age, city):
        self.name = name
        self.age = age
        self.city = city
        self.age_group = self.categorize_age()

    def categorize_age(self):
        """Classify age into 'minner', 'mager', or 'sonner'."""
        if self.age < 20:
            return "minner"
        elif 20 <= self.age < 40:
            return "mager"
        else:
            return "sonner"

class AgeClassifier:
    """Classifies people into age groups and organizes them."""
    def __init__(self, people):
        self.people = [Person(**p) for p in people]

    def group_by_age(self):
        """Group people by their age category."""
        grouped = {}
        for person in self.people:
            if person.age_group not in grouped:
                grouped[person.age_group] = []
            grouped[person.age_group].append(person.name)
        return grouped

# Sample data
people_data = [
    {"name": "Shrish", "age": 18, "city": "Bangalore"},
    {"name": "Hari", "age": 25, "city": "Mumbai"},
    {"name": "Amit", "age": 42, "city": "Delhi"},
    {"name": "Rahul", "age": 35, "city": "Pune"},
    {"name": "Ravi", "age": 12, "city": "Chennai"},
    {"name": "Vikas", "age": 50, "city": "Kolkata"}
]

# Create an instance of AgeClassifier
classifier = AgeClassifier(people_data)

# Get grouped data
result = classifier.group_by_age()

# Print the result 
print(result)


{'minner': ['Shrish', 'Ravi'], 'mager': ['Hari', 'Rahul'], 'sonner': ['Amit', 'Vikas']}


In [1]:
def quicksort(arr):
  if len(arr) < 1:
    return arr 
  else:
    povit = arr[len(arr) //2 ]
    left = [x for x in arr if x < povit]
    right = [x for x in arr if x > povit] 
    mid = [x for x in arr if x == povit]
    return quicksort(left) + mid + quicksort(right)

arr = [3, 6, 8, 10, 1, 2, 1]

print(quicksort(arr))
    

def merge_sort(arr):
    
    if len(arr) <= 1:
        return arr  
    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])  
    right_half = merge_sort(arr[mid:]) 

[1, 1, 2, 3, 6, 8, 10]


In [ ]:
def valid_email(email):
  
  email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\[a-zA-Z]{3,}$'
  if re.match(email_pattern, email):
    return True 
  else:
    return False 
    

In [ ]:
def count_word_func():
  try:
    with open('file.txt','r') as file:
      content  = file.read()
      word = len(content.split())
      return word
  except fileNotFoundError:
    print(f'this file at {file_path} was not found .')
    return  0 
  except exception as e:
    print(f'file error  occurresd: {e}')
    return 0 
    

In [ ]:
def  is_valid(s):
  stack = []
  mapping  ={
    '}':'{',
    ']':'[',
    ')':'('
  }
  
  for char  in s :
    if char  in mapping:
      top_element = stock.pop() if stack else '#'
      if mapping[char] != top_element:
        return False
      else:
        stock.append(char)
  return not stock 
print(is_valid("()[]{}"))  

In [2]:
l1 = [1,2,3,4,5]

l2 = [1,2,4,5,6]
# out= [1,1,2,2,3,4,4,5,5,6]
def marge_two_list(l1,l2):
  
  res =[]
  i,j = 0,0
  while i < len(l1) and j < len(l2):
   
    if l1[i]< l2[j]:
      res.append(l1[i])
      i+=1 
    elif l1[i] > l2[j] :
      res.append(l2[j])
      j+= 1
    else:
      res.append(l1[i])
      res.append(l2[j]) 
      i+=1
      j+=1
  while i < len(l1):
      res.append(l1[i])
      i+= 1
      
  while j <len(l2):
      res.append(l2[j])
      j += 1
  return res 
  
print(marge_two_list(l1,l2))

[1, 1, 2, 2, 3, 4, 4, 5, 5, 6]


In [ ]:
# Given an array of integers, return indices of the two numbers 
# such that they add up to a specific target
l1 = [2, 7, 11, 15]
target = 9

def two_sum(l1, target):
  mepped ={}
  
  for i , num in enumerate(l1):
    diff = target - num 
    if diff in mepped:
      return mepped[diff] , i
    mepped[num] = i 
    
print(two_sum (l1,target)) 

In [ ]:
def make_a_single_list(l1):
    one_list = []
    
    for i in l1:
        if isinstance(i, list):
            one_list.extend(make_a_single_list(i))  # Recursively flatten the list
        elif isinstance(i, dict):
            one_list.extend(i.values())  # Extract values from the dictionary
            one_list.extend(value for value in i.values() if isinstance(value, list) for item in make_a_single_list(value))  # Flatten any lists in the dictionary
        else:
            one_list.append(i)  # Append the item if it's not a list or dictionary
            
    return one_list 

l1 = [[1, 2, 3], [2], [2], [2], [[[22], {'name': 'shrish', 'age': 31, 'courses': ['a', 'b', 'c']}]]]
flattened_list = make_a_single_list(l1)
print(flattened_list)
print(make_a_single_list(l1))

In [ ]:
data = [(1, 3), (2, 1), (3, 2)]
# Expected Output: [(2, 1), (3, 2), (1, 3)]     

result = sorted(data ,key = lambda x :x[0])
print(result)

In [ ]:
def decorator_func(func):
  def warepper_fun(*args,**kwargs):
    print ('before warepper_fun ')
    import time 
    start_time = time.time()
    res =func(*args,**kwargs)
    end_time =time.time()
    diff =end_time - start_time
    return f'{func.__name__} name  {diff} second '


  return warepper_fun 


@decorator_func
def hello_print():
  return 'welcome  hello_print'
  
print(hello_print())

In [3]:
def generater_fun(n):
  a, b =0 , 1
  for i in range(n):
    yield a 
    a, b= b, a+b 
    
n =10
for num in generater_fun(n):
  print(num)


0
1
1
2
3
5
8
13
21
34


In [ ]:
def count_word_func():
  try:
    with open('file.txt','r') as file:
      content  = file.read()
      word = len(content.split())
      return word
  except fileNotFoundError:
    print(f'this file at {file_path} was not found .')
    return  0 
  except exception as e:
    print(f'file error  occurresd: {e}')
    return 0 

In [ ]:
import re 

def valid_email(email):
  
  email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\[a-zA-Z]{3,}$'
  if re.match(email_pattern, email):
    return True 
  else:
    return False

In [ ]:
def quicksort(arr):
  if len(arr) < 1:
    return arr 
  else:
    povit = arr[len(arr) //2 ]
    left = [x for x in arr if x < povit]
    right = [x for x in arr if x > povit] 
    mid = [x for x in arr if x == povit]
    return quicksort(left) + mid + quicksort(right)

arr = [3, 6, 8, 10, 1, 2, 1]

print(quicksort(arr))

In [ ]:
def merge_sort(arr):
    
    if len(arr) <= 1:
        return arr  
    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])  
    right_half = merge_sort(arr[mid:])  

    return merge(left_half, right_half)

def merge(left, right):
   
    sorted_list = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            sorted_list.append(left[i])
            i += 1
        else:
            sorted_list.append(right[j])
            j += 1

    sorted_list.extend(left[i:])
    sorted_list.extend(right[j:])

    return sorted_list

unsorted_list = [38, 27, 43, 3, 9, 82, 10]
sorted_list = merge_sort(unsorted_list)
print("Sorted List:", sorted_list)
      